ibis/pandas/core.py

```diff
@@ -226,9 +245,16 @@ def execute_until_in_scope(
                 op, op.value, expr.type(), aggcontext=aggcontext, **kwargs
             )
         }
-
+    # pre_executed_states is a list of states with same the length of
+    # computable_args, these states are passed to each arg
+    arg_timecontexts = compute_time_context(op, timecontext=timecontext)
     pre_executed_scope = pre_execute(
-        op, *clients, scope=scope, aggcontext=aggcontext, **kwargs
+        op,
+        *clients,
+        scope=scope,
+        timecontext=timecontext,
```

Comment:
How is arg_timecontexts passed to pre_execute here?

Also, can you add a test to excise the pre_execute path?

Reply:
arg_timecontexts computes the time contexts for child nodes, this pre_execute is pre_executing the current node so original timecontext is passed here. But it is a good Q if we want context from children nodes in some custom pre_execute

Reply:
+1 Will add a test to test to path

Reply:
I see. So even though `arg_timecontexts` is already computed, `pre_execute` still needs to recompute it?

Reply:
The ``timecontext`` param here is computed by the parent node of the current node, and ``pre_execute`` will use this time context since it is already computed. ``arg_timecontexts`` will be passed to the for loop later in ``execute_until_in_scope`` and then passed to children nodes, used by ``pre_execute`` of children nodes

Reply:
I see. To be clear, `pre_execute` does not need to compute `arg_timecontexts`?

Reply:
No, it doesn't. We have separate time context computation from pre_execute. It is passed in just in case we needed in some custom ``pre_execute``

Reply:
Ok, sounds good